velox/experimental/cudf/exec/ToCudf.cpp

```diff
@@ -0,0 +1,219 @@
+
+    bool const kPreviousOperatorIsNotGpu =
+        (operatorIndex > 0 and !isSupportedGpuOperators[operatorIndex - 1]);
+    bool const kNextOperatorIsNotGpu =
+        (operatorIndex < operators.size() - 1 and
+         !isSupportedGpuOperators[operatorIndex + 1]);
+
+    auto id = oper->operatorId();
+    if (kPreviousOperatorIsNotGpu and acceptsGpuInput(oper)) {
+      auto planNode = getPlanNode(oper->planNodeId());
+      replaceOp.push_back(std::make_unique<CudfFromVelox>(
+          id, planNode->outputType(), ctx, planNode->id() + "-from-velox"));
+      replaceOp.back()->initialize();
+    }
+
+    if (auto orderByOp = dynamic_cast<exec::OrderBy*>(oper)) {
+      auto id = orderByOp->operatorId();
+      auto planNode = std::dynamic_pointer_cast<const core::OrderByNode>(
+          getPlanNode(orderByOp->planNodeId()));
+      VELOX_CHECK(planNode != nullptr);
+      replaceOp.push_back(std::make_unique<CudfOrderBy>(id, ctx, planNode));
+      replaceOp.back()->initialize();
+    }
+
+    if (kNextOperatorIsNotGpu and producesGpuOutput(oper)) {
+      auto planNode = getPlanNode(oper->planNodeId());
+      replaceOp.push_back(std::make_unique<CudfToVelox>(
+          id, planNode->outputType(), ctx, planNode->id() + "-to-velox"));
+      replaceOp.back()->initialize();
+    }
+
+    if (not replaceOp.empty()) {
+      operatorsOffset += replaceOp.size() - 1;
+      [[maybe_unused]] auto replaced = driverFactory_.replaceOperators(
+          driver_,
+          replacingOperatorIndex,
+          replacingOperatorIndex + 1,
+          std::move(replaceOp));
+      replacementsMade = true;
+    }
+  }
+
+  return replacementsMade;
+}
+
+struct CudfDriverAdapter {
+  std::shared_ptr<rmm::mr::device_memory_resource> mr_;
+  std::shared_ptr<std::vector<core::PlanNodePtr>> planNodes_;
+
+  CudfDriverAdapter(std::shared_ptr<rmm::mr::device_memory_resource> mr)
+      : mr_(mr) {
+    planNodes_ = std::make_shared<std::vector<core::PlanNodePtr>>();
+  }
+
+  // Call operator needed by DriverAdapter
+  bool operator()(const exec::DriverFactory& factory, exec::Driver& driver) {
+    auto state = CompileState(factory, driver, *planNodes_);
+    // Stored planNodes_ from inspect.
+    auto res = state.compile();
+    return res;
+  }
+
+  // Iterate recursively and store them in the planNodes_.
+  void storePlanNodes(const core::PlanNodePtr& planNode) {
+    const auto& sources = planNode->sources();
+    for (int32_t i = 0; i < sources.size(); ++i) {
+      storePlanNodes(sources[i]);
+    }
+    planNodes_->push_back(planNode);
+  }
+
+  // Call operator needed by plan inspection
+  void operator()(const core::PlanFragment& planFragment) {
+    // signature: std::function<void(const core::PlanFragment&)> inspect;
+    // call: adapter.inspect(planFragment);
+    planNodes_->clear();
+    if (planNodes_) {
+      storePlanNodes(planFragment.planNode);
+    }
+  }
+};
+
+static bool isCudfRegistered = false;
+
+void registerCudf() {
```

Comment:
registerCudf may can accept argument CudfOptions to set VELOX_CUDF_DISABLED and VELOX_CUDF_MEMORY_RESOURCE

Reply:
With the suggested change, we'd have to add a couple of env var checks everywhere we try to call `registerCudf`. I'm not sure if that's the intended way to use `registerCudf`. In fact, I'm not sure where the calls to register these driver adapters are supposed to be in non-test code, and if those locations have access to some global options. @karthikeyann can you chime in?

Reply:
I try to integrate cudf in project Gluten, Spark is a distributed system, if we need to set env, it will be not convenient.

Reply:
@jinchengchenghh What's the preferred place to put these options? Is it QueryConfig?

Reply:
No, ToCudf.h is enough. I will directly call `registerCudf`.

Reply:
Will this be suitable?
gflags as default, but can override with argument to `registerCudf`
```cpp
DEFINE_bool(velox_cudf_enabled, true, "Enable cuDF-Velox acceleration");
DEFINE_string(velox_cudf_memory_resource, "async", "Memory resource for cuDF");
```

```cpp
#include <gflags/gflags.h>

DECLARE_bool(velox_cudf_enabled);
DECLARE_string(velox_cudf_memory_resource);

struct CudfOptions {
  bool cudfEnabled = FLAGS_velox_cudf_enabled;
  std::string cudfMemoryResource = FLAGS_velox_cudf_memory_resource;
  static CudfOptions defaultOptions() {
    return CudfOptions();
  }
};

/// Registers adapter to add cuDF operators to Drivers.
void registerCudf(const CudfOptions& options = CudfOptions::defaultOptions());
```

Reply:
I only need the CudfOptions, I don't thin we need the gflags.
Maybe
```
struct CudfOptions {
  bool cudfEnabled = true;
  std::string cudfMemoryResource = "async";
  static CudfOptions defaultOptions() {
    return CudfOptions();
  }
};
```

Reply:
If it's not much of a code smell, we'd like to keep the gflags so we can conveniently run benchmarks with and without cudf adaptation. Although maybe we can pull the `cudfEnabled` out of the options as we likely shouldn't call `registerCudf()` if we don't want it running. Does this align with how you'll use `registerCudf()` @jinchengchenghh ?

Reply:
Yes, sounds good.